### PR TITLE
Add deepspeed and profiler configurations

### PR DESCRIPTION
**Pull request by lscheinkman**
Update deepspeed configurations compatible with deepspeed 0.4.0 using deepspeed inference engine for KD teacher model

### CODE REVIEW

File: projects/transformers/trainer_mixins/distillation.py [hunk 1]
```diff
@@ -103,9 +109,16 @@ def __init__(self, *args, **kwargs):
...
                 cache_dir=teacher_models_cache_dir
             )
-            if self.args.fp16:
-                teacher_model.half()
             teacher_model.resize_token_embeddings(len(self.tokenizer))
+
+            # Use deepspeed inference mode on teacher models
+            if self.args.deepspeed:
+                ds_engine = deepspeed.init_inference(
+                    teacher_model, dtype=torch.half, replace_method="auto")
```
Comment (lucasosouza) [inline projects/transformers/trainer_mixins/distillation.py:117]: This is inference only, also might not make a big difference in speed if we are using 16 or 32 float (unless 16 is a requirement for deepspeed). If there is an issue with acc we can try using float 32 instead.

**Review by lucasosouza — APPROVED** [review]
Looks good, left a comment on whether or not we should use fp16 for inference on the teacher model .

File: projects/transformers/trainer_mixins/distillation.py (hunk 1 above)
Comment (lscheinkman) [reply projects/transformers/trainer_mixins/distillation.py:117]: I will try that and update the PR if it helps

Comment (lscheinkman) [reply projects/transformers/trainer_mixins/distillation.py:117]: When i used fp32 deepspeed crashed with `CUDA error: an illegal memory access was encountered`. See trace below
```
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/transformers/models/bert/modeling_bert.py", line 1320, in forward                                                         
    return_dict=return_dict,                                                                                                                                  
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl       
    result = self.forward(*input, **kwargs)                                                                                           
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/transformers/models/bert/modeling_bert.py", line 981, in forward                                                                     
    return_dict=return_dict,                                                                                                                                 
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl                                   
    result = self.forward(*input, **kwargs)                                                                                                                                                                     
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/transformers/models/bert/modeling_bert.py", line 575, in forward                                                                     
    output_attentions,                                                                                                                                                                                          
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl                                                                                 
    result = self.forward(*input, **kwargs)                                                                                                                                                                     
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 611, in forward         
    self.norm_b)                                                                                                                                                                                                
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl                                                                                 
    result = self.forward(*input, **kwargs)                                                                                                                   
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 393, in forward                                                  
    self.qkv_merging)                                                                                                                                                                                           
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 308, in forward    
    output, key_layer, value_layer, context_layer = selfAttention_fp()                                                                                                                                          
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 273, in selfAttention_fp                                         
    context_layer, key_layer, value_layer = compute_attention(qkv_out)                                                                                                                    
  File "/home/ubuntu/miniconda3/envs/deepspeed/lib/python3.7/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 167, in compute_attention                                        
    qkv_out = qkv_out.float()                                                                                                                                                                                   
RuntimeError: CUDA error: an illegal memory access was encountered  
```

**Pull request merged**